File: 530.py
"""
Use inorder traversal to get a list of values then calculate min diff.
Actually, we only need to calculate it without get the whole list.
i.e.
self.lastNode = root.val
root.val - self.lastnode
"""
class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        values = self.inorderTraversal(root)
        minDiff = float('inf')
        for i in range(len(values)-1):
            minDiff = min(minDiff, values[i+1] - values[i])
        return minDiff
        
    def inorderTraversal(self, root):
        left, right = [], []
        if root.left:
            left = self.inorderTraversal(root.left)
        if root.right:
            right = self.inorderTraversal(root.right)
        return left + [root.val] + right

File: go/530.go
/*
Didn't read the question carefully. The tree is BST actually, so no need to sort if push all elements to left of the array.
*/
func getMinimumDifference(root *TreeNode) int {
    nums := make([]int, 0)
    helper(root, &nums)
    sort.Ints(nums)
    minDiff := math.MaxInt32
    for i := 0; i < len(nums) - 1; i++ {
        diff := nums[i+1] - nums[i]
        if diff < minDiff {
            minDiff = diff
        }
    }
    return minDiff
}

func helper(root *TreeNode, nums *[]int) {
    if root == nil {
        return
    }
    *nums = append(*nums, root.Val)
    if root.Left != nil {
        helper(root.Left, nums)
    }
    if root.Right != nil {
        helper(root.Right, nums)
    }
    return
}
